### PR TITLE
Ensure a redirect of the HttpServletResponse will close HttpOutput

### PR DESCRIPTION
**Pull request by lachlan-roberts**
In the GAE issue https://github.com/GoogleCloudPlatform/appengine-java-standard/issues/321, a redirect is being done from a JSP file with non-empty content.

In Jetty 9.4 and EE8 we would close the `HttpOutput` during the call to `sendRedirect`, and so any further content that is tried to send should throw `IOException`.

### CODE REVIEW

**Review by sbordet — DISMISSED** [review]

File: jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletApiResponse.java [hunk 1]
```diff
@@ -186,9 +186,19 @@ public void sendRedirect(int code, String location) throws IOException
         {
             Response.sendRedirect(getServletRequestInfo().getRequest(), getResponse(), callback, code, location, false);
             callback.block();
+            closeOutput();
```
Comment (gregw) [inline jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletApiResponse.java]: Why is this necessary? The send redirect above ultimately does
```java
            response.getHeaders().put(HttpHeader.LOCATION, toRedirectURI(request, location));
            response.setStatus(code);
            response.write(true, null, callback);
```
So the last write has been done and no more content can be sent.
Is this purely to make sure an IOE is thrown even when aggregating into the HttpOutput buffer?   If so, then rather than a new method how about:
```suggestion
            getServletChannel().getHttpOutput().close();
```

**Review by gregw — CHANGES_REQUESTED** [review]

File: jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletApiResponse.java (hunk 1 above)
Comment (lachlan-roberts) [reply jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletApiResponse.java]: > Is this purely to make sure an IOE is thrown even when aggregating into the HttpOutput buffer?

Yes, exactly. So without this, writes from a Servlet may not throw as they are just aggregated, and it only throws when it does the complete in ServletChannel. In GAE this was causing a 500 response instead of the proper redirect code to be sent if a redirect was done from a JSP.

How about I inline this method then or make it private, because I still think we need to close the writer if there is one.

Comment (gregw) [reply jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletApiResponse.java]: OK

**Comment by lachlan-roberts** [comment]
@gregw please re-review

**Review by gregw — APPROVED** [review]

**Comment by MuffinTheMan** [comment]
What are the chances this makes the next release? Also, @lachlan-roberts, how soon after release will this make it into App Engine?

**Comment by lachlan-roberts** [comment]
@MuffinTheMan This will make it into the next release of `12.0.17`, we try to release at the end of every month, so this will likely be late Dec or very early Jan.

After this is released I will update GAE to `12.0.17`, but I do not know how long it will take Google to release GAE after that.

You could build Jetty and GAE yourself and deploy your application with this fix if you need it earlier.

**Comment by MuffinTheMan** [comment]
Checking in since it's been a couple months since the most recent release--when is `12.0.17` getting released? It's holding up our upgrade to EE10 😬

**Comment by gregw** [comment]
The release should be this week

**Comment by MuffinTheMan** [comment]
@lachlan-roberts, I'm not sure who handles bumping the version for GAE, but I just saw `12.0.17` was released!

**Comment by lachlan-roberts** [comment]
@MuffinTheMan  I have already updated GAE to 12.0.17 with PR https://github.com/GoogleCloudPlatform/appengine-java-standard/pull/348, but I am not sure when they push this into prod.
